lib_staging.ks

```diff
@@ -1,51 +1,96 @@
+// https://en.wikipedia.org/wiki/Specific_impulse
+// https://en.wikipedia.org/wiki/Standard_gravity
+global isp_g0 is kerbin:mu/kerbin:radius^2. // exactly 9.81 in KSP 1.3.1, 9.80665 for Earth
+// note that constang:G*kerbin:mass/kerbin:radius^2 yields 9.80964723...
+
+// work variables for staging logic
+global stagingNumber	is -1.		// stage:number when last calling stagingPrepare()
+global stagingEngines	is list().	// list of engines that all need to flameout to stage
+global stagingTanks		is list().	// list of tanks that all need to be empty to stage
+// info for and from stageDeltaV
+global stageAvgIsp		is 0.		// average ISP in seconds
+global stageStdIsp		is 0.		// average ISP in N*s/kg (stageAvgIsp*isp_g0)
+global stageDryMass		is 0.		// dry mass just before staging
+global stageBurnTime	is 0.		// updated in stageDeltaV()
+
+// return stage number where the part is decoupled (needed for engines)
+function stagingDecoupledIn {
+	parameter part.
+	// for all parts except engines, :STAGE is the stage number where they are decoupled
+	// for engines, this is the number they are activated, so we find first non-engine part in parent list
+	until not part:istype("engine") {
+		if not part:hasParent return -1. // cannot separate root
+		set part to part:parent.
+	}
+	return part:stage.
+}
+
+// to be called whenever current stage changes to prepare data for quicker test and other functions
 function stagingPrepare {
 	wait until stage:ready.
 	set stagingNumber	to stage:number.
 	set stagingEngines	to list().
 	set stagingTanks	to list().
 	if stage:number = 0 return.
 
+	// prepare list of engines that are to be decoupled by staging
 	list engines in engines.
-	for e in engines if e:stage = stage:number and e:parent:stage = stage:number-1
+	for e in engines if e:stage = stage:number and stagingDecoupledIn(e) = stage:number-1
 		stagingEngines:add(e).
 
-	// prepare list of tanks that are in current stage but not in next stage
+	// prepare list of tanks that are to be decoupled and have some fuel
 	list parts in parts.
 	for t in parts if t:stage = stage:number-1 {
 		local amount is 0.
-		for r in t:resources if stagingFuelList:contains(r:name)
+		for r in t:resources if stagingTankFuels:contains(r:name)
 			set amount to amount + r:amount.
 		if amount > 0.01
 			stagingTanks:add(t).
 	}
+
+	// prepare average ISP for stageDeltaV()
+	local thrust is 0.
+    local flow is 0.
+	list engines in engines.
+	for e in engines if e:ignition and e:isp > 0 {
+		local t is e:availableThrust.
+		set thrust to thrust + t.
+		set flow to flow + t / e:isp. // thrust=isp*g0*dm/dt => flow = sum of thrust/isp
+	}
+	set stageAvgIsp to 0.
+    if flow > 0 set stageAvgIsp to thrust/flow.
+	set stageStdIsp to stageAvgIsp * isp_g0.
+
+	// prepare dry mass for stageDeltaV()
+    local fuelMass is 0.
+    for r in stage:resources if stagingConsumed:contains(r:name)
+		set fuelMass to fuelMass + r:amount*r:density.
+	set stageDryMass to ship:mass-fuelMass.
 }
```

Comment:
I hope to be able to add `nextStageMass`, dry mass and avg.+std. isp to be able to use that in `burnTimeForDv` down there to predict the burn time even if you need to stage during the burn. Current and next should be enough.

Alternative would be to somehow calculate the necessary info for all the stages ahead of time (e.g. in launch script) and store it localy in the kOS CPU.

Reply:
The only way to actually do it in current kOS would be to create some database of engines (and boosters). Maybe even create some script to manage this - create test rocket with the engine/booster and run the script, it will ignite, measure and update the database.

Reply:
Meh; building a database this seems like a lot of work when the alternative is just to tailor your fuel load for clean staging between burns.